plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs

```diff
@@ -36,15 +36,15 @@ instance Alt (Contract i o) where
 -- The applicative instance parallelises the 'Waiting' operations
 instance Applicative (Contract i o) where
     pure = Pure
-    cf <*> ca = case cf of
-        Emit t c -> Emit t (c <*> ca)
-        Pure f -> fmap f ca
-        Waiting f ->
-            case ca of
-                Emit t c -> Emit t (Waiting f <*> c)
-                Pure a' -> Waiting $ \i' -> fmap (\f' -> f' a') (f i')
-                Waiting f' ->
-                    Waiting $ \i' -> f i' <*> f' i'
+    Emit t c  <*> ca       = Emit t (c <*> ca)
+    cf        <*> Emit t c = Emit t (cf <*> c)
+    Pure f    <*> ca       = f <$> ca
+    Waiting f <*> ca       = Waiting $ \i -> f i <*> offer i ca
+
+offer :: i -> Contract i o a -> Contract i o a
```

Comment:
fwiw I think applyInput is a fine name, I just picked this not to clash :laughing: up to you